hasktorch/src/Torch/Data/Pipeline.hs

```diff
@@ -11,145 +14,127 @@
+
+data MapStyleOptions = MapStyleOptions { bufferSize :: Int
+                                       , numWorkers :: Int
+                                       , shuffle :: Sample
+                                       }
+
+mapStyleOpts numWorkers = MapStyleOptions { bufferSize = numWorkers
+                                          , numWorkers = numWorkers
+                                          , shuffle = Sequential
+                                          }
+
+data Sample = Sequential | Shuffle StdGen
+
+class (Ord k) => Dataset m dataset k sample | dataset -> sample, dataset -> k  where
+  getItem :: dataset -> k -> m sample
+  keys :: dataset -> Set k
+
+---------------------- Workflow --------------------
+-- - make a new map of keys to TVars of samples, possibly shuffled keys, tracking which keys have been sampled
+-- - create a TQueue of keys (using pipes-concurrency wrapper)
+-- - fork off workers which all pull from the TQueue and sample that key using the dataset,
+--   then update the TVar associated with that key
+-- have a worker waiting for each successive key to be updated in the list of (key, TVar)
+
+makeListT :: forall m dataset k sample r .
+  (MonadIO m, MonadBaseControl IO m,  Dataset m dataset k sample) =>
+  MapStyleOptions ->
+  dataset ->
+  ContT r m  (ListT m (sample, Int))
+makeListT MapStyleOptions{..} dataset = do
+  (keyOutput, keyInput, seal) <- liftIO $ spawn' unbounded
+
+  let retreiveSet = liftIO $ keyTVarSet $ keys @m dataset
+  keyTVarSet <- case shuffle of
+    Sequential -> retreiveSet
+    Shuffle g -> fst . fisherYates g <$> retreiveSet
+   
+  -- fill the queue with each key and associated TVar then seal it
+  keyQueue keyOutput $ keyTVarSet
+  liftIO $ atomically seal
+
+  let workers = runWorkers numWorkers dataset keyInput
+      datastream = awaitNextItem keyTVarSet
+  runWithBuffer bufferSize $ \output -> concurrently_ workers (datastream output)
+
+runWorkers ::
+  (Dataset m dataset k sample, MonadIO m, MonadBaseControl IO m) =>
+  Int ->
+  dataset ->
+  Input (k, TVar (Maybe sample)) ->
+  m ()
+runWorkers numWorkers dataset keyInput = replicateConcurrently_ numWorkers (runEffect $ fromInput' keyInput >-> runWorker)
+    where 
+      runWorker = forever $ do (key, tvar) <- await
+                               item <- lift $ getItem dataset key
+                               liftIO $ atomically $ writeTVar tvar (Just item) 
+awaitNextItem ::
+  (MonadBase IO m, MonadIO m) =>
+  [(k, TVar (Maybe sample))] ->
+  Output sample ->
+  m ()
+awaitNextItem tvars output  = runEffect $ each tvars >-> readNextItem >-> toOutput' output
+  where readNextItem  = forever $ do
+          (key, tvar) <- await
+          item <- liftIO $ atomically $ do
+            val <- readTVar tvar 
+            case val of
+              Nothing -> retry
+              Just item -> writeTVar tvar Nothing >> pure item -- reset the tvar once we get the sample out of it to save memory
+          yield item 
```

Comment:
during our call this morning, Adam remarked that it is possible that when one of the workers above stalls, the other workers will continue filling up the tvars, potentially until memory runs out, because the downstream consumer here is blocked.
You said this can be resolved easily, but I am missing this here.
A possible solution would be to restrict the number of keys that are pushed to the `keyInput` in such a way that it can never go beyond some `n` keys after the last unfulfilled tvar.